src/Microsoft.CodeQuality.Analyzers/Core/Maintainability/RemoveUnusedLocals.Fixer.cs

```diff
@@ -1,27 +1,174 @@
     /// CA1804: Remove unused locals
     /// </summary>
+    [ExportCodeFixProvider(LanguageNames.CSharp, LanguageNames.VisualBasic, Name = AvoidUnusedPrivateFieldsAnalyzer.RuleId), Shared]
     public abstract class RemoveUnusedLocalsFixer : CodeFixProvider
     {
-        public sealed override FixAllProvider GetFixAllProvider()
+        internal const string RuleId = "CA1804";
+        private readonly NodesProvider _nodesProvider;
+
+        protected RemoveUnusedLocalsFixer(NodesProvider nodesProvider) => _nodesProvider = nodesProvider;
+
+        public sealed override FixAllProvider GetFixAllProvider() => new RemoveLocalFixAllProvider(_nodesProvider);
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Diagnostic diagnostic = context.Diagnostics.Single();
+            var nodesToRemove = await _nodesProvider.GetNodesToRemoveAsync(context.Document, diagnostic, context.CancellationToken).ConfigureAwait(false);
+
+            context.RegisterCodeFix(
+                new RemoveLocalAction(
+                    MicrosoftMaintainabilityAnalyzersResources.RemoveUnusedLocalsMessage,
+                    async ct => await _nodesProvider.RemoveNodes(context.Document, nodesToRemove, ct).ConfigureAwait(false)),
+                diagnostic);
+        }
+
+        protected abstract class NodesProvider
+        {
+            public abstract SyntaxNode GetNodeToRemoveOrReplace(SyntaxNode node);
+
+            public abstract void RemoveAllUnusedLocalDeclarations(HashSet<SyntaxNode> nodesToRemove);
+
+            public async Task<ImmutableArray<SyntaxNode>> GetNodesToRemoveAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+            {
+                SyntaxNode root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                SyntaxNode node = root.FindNode(diagnostic.Location.SourceSpan);
+
+                DocumentEditor editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+                ISymbol symbol = editor.SemanticModel.GetDeclaredSymbol(node);
+                var referencedSymbols = await SymbolFinder.FindReferencesAsync(symbol, document.Project.Solution, cancellationToken).ConfigureAwait(false);
+
+                var nodesToRemoveBuilder = ImmutableArray.CreateBuilder<SyntaxNode>();
+                foreach (var referencedSymbol in referencedSymbols)
+                {
+                    if (referencedSymbol?.Locations != null)
+                    {
+                        foreach (var location in referencedSymbol.Locations)
+                        {
+                            var referencedSymbolNode = root.FindNode(location.Location.SourceSpan);
+                            if (referencedSymbolNode != null)
+                            {
+                                var assignmentNode = GetNodeToRemoveOrReplace(referencedSymbolNode);
```

Comment:
Is this always an assignment node? Can you use explicit type name instead of var for clarity?